packages/flutter/lib/src/widgets/router.dart

```diff
@@ -0,0 +1,1012 @@
+  /// the current configuration.
+  ///
+  /// At most one [Router] can opt in to route name reporting. In common use
+  /// case, only the top-most [Router] created by [WidgetsApp.router] should
+  /// opt for route name reporting.
+  T get currentConfiguration => null;
+
+  /// Discards any resources used by the object.
+  ///
+  /// After this is called, the object is not in a usable state and should be
+  /// discarded.
+  ///
+  /// This method should only be called by the object's owner.
+  @mustCallSuper
+  void dispose() { }
+
+  /// Called by the [Router] to obtain the widget tree that represents the
+  /// current state.
+  ///
+  /// This is called whenever the [setInitialRoutePath] method's future
+  /// completes, the [setNewRoutePath] method's future completes with the value
+  /// true, the [popRoute] method's future completes with the value true, or
+  /// this object notifies its clients (see the [Listenable] interface, which
+  /// this interface includes). In addition, it may be called at other times. It
+  /// is important, therefore, that the methods above do not update the state
+  /// that the [build] method uses before they complete their respective
+  /// futures.
+  ///
+  /// Typically this method returns a suitably-configured [Navigator]. If you do
+  /// plan to create a navigator, consider using the
+  /// [PopNavigatorRouterDelegateMixin].
+  ///
+  /// This method must not return null.
+  ///
+  /// The `context` is the [Router]'s build context.
+  Widget build(BuildContext context);
+}
+
+/// A route name provider that provides route name for the [Router] widget
+///
+/// This provider is responsible for handing the route name through [value]
+/// getter and notifies listeners, typically the [Router] widget, when a new
+/// route name is available.
+///
+/// If you decide to opt for the route name reporting (by overrides the
+/// [RouterDelegate.currentConfiguration] to return non-null), overrides the
+/// [routerReportsNewRouteName] method to reports the route name back to
+/// the engine or uses the [RouteNameProviderReportBack] mixin to do so.
+///
+/// See also:
+///
+///  * [PlatformRouteNameProvider], which wires up the provider with the
+///    [WidgetsBindingObserver.didPushRoute] to propagate platform push route
+///    intent to the [Router] widget.
+///  * [RouteNameProviderReportBack], which provides the implementation of
+///    [routerReportsNewRouteName].
+abstract class RouteNameProvider extends ValueListenable<String> {
+  /// A callback called when the [Router] widget has a new route name that is
+  /// different from the existing route name in this provider.
+  void routerReportsNewRouteName(String routeName) { }
+
+  /// Discards any resources used by the object.
+  ///
+  /// After this is called, the object is not in a usable state and should be
+  /// discarded.
+  ///
+  /// This method should only be called by the object's owner.
+  @mustCallSuper
+  void dispose() { }
```

Comment:
@Hixie 
I am wondering which makes more senses
1, creates dispose method in base class for all components (RouteNameParser, RouteNameProvider, and RouterDelegate, BackButtonDispatcher)
2, only create dispose method in component subclasses that are relevant(PlatformRouteNameProvider, or StockRouterDelegate in the stock app)
3, create dispose method in base class that we care(for example we don't provide a RouteNameParser subclass that will need to release resource)

Reply:
You should put a dispose method in the base class of hierarchies that have a clear ownership model where the objects will not live longer than their owner and where the classes may have mutable state or may allocate resources. Generally, anything that has listeners, for example, likely needs to have a clear ownership model and thus a dispose method. See also the discussion in our style guide.